# Data Preprocessing Assignment

Dataset used: Titanic Passenger List


## Importing Modules and Reading the Data

In [1]:
import pandas as pd
import numpy as np

In [37]:
train_data = pd.read_csv("train.csv")
test_data = pd.read_csv("test.csv")

In [38]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [39]:
test_data.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [40]:
train_data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [41]:
test_data.columns

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [42]:
print("Length of train data: " + str(len(train_data)))
print("Length of test data: " + str(len(test_data)))

Length of train data: 899
Length of test data: 418


## Removing Duplicate Values


### Checking For & Removing Duplicates

In [43]:
dups = train_data.duplicated()
print(dups.any())

True


In [44]:
train_data = train_data.drop_duplicates()
print("Length of Train data after removing duplicates: " + str(len(train_data)))

Length of Train data after removing duplicates: 891


In [45]:
dups = test_data.duplicated()
print(dups.any())

False


## Analysing Data

In [46]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 894
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 90.5+ KB


In [47]:
train_data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Looking at the data given above, we can find that 38% of the people in the train data survived. Since the number of records in the dataset is 891, we can conclude that there are missing values in the column "Age".

### Finding Missing Data

In [48]:
missing = train_data.isnull().sum().sort_values(ascending=False)
percent_1 = train_data.isnull().sum()/train_data.isnull().count()*100
percent_2 = (round(percent_1, 1)).sort_values(ascending=False)
missing_data = pd.concat([missing, percent_2], axis=1, keys=['Total', '%'])

missing_data

Total     %
Cabin          687  77.1
Age            177  19.9
Embarked         2   0.2
PassengerId      0   0.0
Survived         0   0.0
Pclass           0   0.0
Name             0   0.0
Sex              0   0.0
SibSp            0   0.0
Parch            0   0.0
Ticket           0   0.0
Fare             0   0.0

From the data above, we can conclude that there are data missing in the columns "Cabin", "Age" and "Embarked". \
The number of missing value in "Cabin" is very large (77.1%), so we could remove this column. \
The rest of the missing values, i.e., for "Age" and "Embarked", should be filled appropriately.

In [66]:
missing = test_data.isnull().sum().sort_values(ascending=False)
percent_1 = test_data.isnull().sum()/test_data.isnull().count()*100
percent_2 = (round(percent_1, 1)).sort_values(ascending=False)
missing_data = pd.concat([missing, percent_2], axis=1, keys=['Total', '%'])

missing_data

Total     %
Age          86  20.6
Pclass        0   0.0
Sex           0   0.0
SibSp         0   0.0
Parch         0   0.0
Embarked      0   0.0

In the test data, there are missing values for "Age". These are to be filled appropriately.

### Finding Columns that doesn't attribute to the result

In [49]:
train_data.columns.values

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

In [50]:
test_data.columns.values

array(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

It would be fine to assume that 'Name', 'PassengerId', 'Ticket', 'Fare' doesn't attribute to the result. So they could also be removed.

## Deleting Columns

Analysing the data, we concluded that the columns "Name", "PassengerId", "Ticket", "Cabin", "Fare" could be removed.

In [51]:
del train_data["Name"]
del train_data["PassengerId"]
del train_data["Fare"]
del train_data["Ticket"]
del train_data["Cabin"]

train_data.head()

Survived  Pclass     Sex   Age  SibSp  Parch Embarked
0         0       3    male  22.0      1      0        S
1         1       1  female  38.0      1      0        C
2         1       3  female  26.0      0      0        S
3         1       1  female  35.0      1      0        S
4         0       3    male  35.0      0      0        S

In [53]:
del test_data["Name"]
del test_data["PassengerId"]
del test_data["Fare"]
del test_data["Ticket"]
del test_data["Cabin"]

test_data.head()

Pclass     Sex   Age  SibSp  Parch Embarked
0       3    male  34.5      0      0        Q
1       3  female  47.0      1      0        S
2       2    male  62.0      0      0        Q
3       3    male  27.0      0      0        S
4       3  female  22.0      1      1        S

## Filling Missing Data

The columns "Age" and "Embarked" had missing values. These are to filled in appropriately.

### Filling missing "Age"

As seen in the description of the data, the value of age varies from 0.4 to 80. It would make sense to fill the missing values with the mean of the column. \
For those who survived and has a missing age value, those can be filled with the mean age of those survived. Similarly, mean age of those who didn't survive can be used to fill the age value of those who have age missing and didn't survive. \
For test data, the missing values can be filled with the mean values.

In [54]:
meanS = train_data[train_data.Survived == 1].Age.mean()
meanNS = train_data[train_data.Survived == 0].Age.mean()

print(meanS)
print(meanNS)

28.343689655172415
30.62617924528302


In [57]:
train_data["Age"] = np.where(pd.isnull(train_data.Age) & train_data["Survived"] == 1, meanS, train_data.Age)
train_data.Age.fillna(meanNS, inplace=True)


In [67]:
mean = test_data.Age.mean()

print(mean)

30.272590361445783


In [68]:
test_data.Age.fillna(mean, inplace=True)

### Filling Missing "Embarked"

In [59]:
train_data.Embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [60]:
print(train_data[train_data.Survived == 1][train_data.Embarked == 'S'].shape[0])
print(train_data[train_data.Survived == 1][train_data.Embarked == 'C'].shape[0])
print(train_data[train_data.Survived == 1][train_data.Embarked == 'Q'].shape[0])

217
93
30


<ipython-input-60-d9cab52c6f4d>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  print(train_data[train_data.Survived == 1][train_data.Embarked == 'S'].shape[0])
<ipython-input-60-d9cab52c6f4d>:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  print(train_data[train_data.Survived == 1][train_data.Embarked == 'C'].shape[0])
<ipython-input-60-d9cab52c6f4d>:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  print(train_data[train_data.Survived == 1][train_data.Embarked == 'Q'].shape[0])


As seen in the data given above, the value of the column "Embarked" does affect survival of an individual. Now, since the number of missing data in this case is 2, and insignificant, we can remove these 2 rows.

In [62]:
# The only nan values are in Embarked. So we can use dropna()
train_data.dropna(inplace=True)

In [63]:
train_data.shape

(889, 7)

### Checking if all the null values have been replaced or removed

In [69]:
train_data.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Embarked    0
dtype: int64

In [70]:
test_data.isnull().sum()

Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Embarked    0
dtype: int64

## Updating Column Values - Converting to Numerical Values

### Updating data in the column "Sex"
Updating the column "Sex" to values 0 & 1, when Sex=male and Sex=female, respectively.

In [64]:
train_data["Sex"] = np.where(train_data["Sex"] == "male", 0, 1)

train_data.head()

Survived  Pclass  Sex   Age  SibSp  Parch Embarked
0         0       3    0  22.0      1      0        S
1         1       1    1  38.0      1      0        C
2         1       3    1  26.0      0      0        S
3         1       1    1  35.0      1      0        S
4         0       3    0  35.0      0      0        S

In [76]:
test_data["Sex"] = np.where(test_data["Sex"] == "male", 0, 1)

test_data.head()

Pclass  Sex   Age  SibSp  Parch Embarked
0       3    0  34.5      0      0        Q
1       3    1  47.0      1      0        S
2       2    0  62.0      0      0        Q
3       3    0  27.0      0      0        S
4       3    1  22.0      1      1        S

### Updating Values in "Embarked"

The values in Embarked are characters. They can be replaced with integers.

In [73]:
train_data["Embarked"].unique()

array(['S', 'C', 'Q'], dtype=object)

In [75]:
train_data.replace(to_replace=['S', 'C', 'Q'], value=[1, 2, 3], inplace=True)

train_data.head()

Survived  Pclass  Sex   Age  SibSp  Parch  Embarked
0         0       3    0  22.0      1      0         1
1         1       1    1  38.0      1      0         2
2         1       3    1  26.0      0      0         1
3         1       1    1  35.0      1      0         1
4         0       3    0  35.0      0      0         1

In [77]:
test_data.replace(to_replace=['S', 'C', 'Q'], value=[1, 2, 3], inplace=True)

test_data.head()

Pclass  Sex   Age  SibSp  Parch  Embarked
0       3    0  34.5      0      0         3
1       3    1  47.0      1      0         1
2       2    0  62.0      0      0         3
3       3    0  27.0      0      0         1
4       3    1  22.0      1      1         1

## Prediction Model

This is a classifier problem. An individual is to be classified as "Survived" or "Not Survived". \
The classifier model which will be used here is "Random Forest".

In [78]:
from sklearn.ensemble import RandomForestClassifier

In [81]:
train_data.head()

Survived  Pclass  Sex   Age  SibSp  Parch  Embarked
0         0       3    0  22.0      1      0         1
1         1       1    1  38.0      1      0         2
2         1       3    1  26.0      0      0         1
3         1       1    1  35.0      1      0         1
4         0       3    0  35.0      0      0         1

In [82]:
y = train_data["Survived"]
features = ["Pclass", "Sex", "SibSp", "Parch", "Age", "Embarked"]
X = pd.get_dummies(train_data[features])
X_test = pd.get_dummies(test_data[features])

In [83]:
X.head()

Pclass  Sex  SibSp  Parch   Age  Embarked
0       3    0      1      0  22.0         1
1       1    1      1      0  38.0         2
2       3    1      0      0  26.0         1
3       1    1      1      0  35.0         1
4       3    0      0      0  35.0         1

In [84]:
model = RandomForestClassifier(n_estimators=100, max_depth=5, random_state=1)
model.fit(X, y)
predictions = model.predict(X_test)

In [88]:
output = pd.DataFrame({'Survived': predictions})
output.head()

Survived
0         0
1         0
2         0
3         0
4         1

## References: 
https://www.kaggle.com/anirban7/data-preprocessing-for-beginners \
https://www.kaggle.com/alexisbcook/titanic-tutorial \
https://towardsdatascience.com/predicting-the-survival-of-titanic-passengers-30870ccc7e8

In [71]:
# %history -n -o 1-27